File: packages/ink-swift/Sources/InkCalligrapher/Engine.swift
/// Generation API over the calligrapher cell, ported from
/// packages/ink-calligrapher/src/engine.ts: encode [START, ...text, END],
/// condition on one of the 80 learned style vectors (or a seed-picked
/// random one), then autoregressively sample until the attention head
/// signals the text is exhausted or the step budget runs out. The
/// terminating step's sample is discarded, exactly like the reference.

import Foundation
import InkCore

public enum CalligrapherError: Error, CustomStringConvertible {
    case missingTensor(String)
    case missingSparseTensor(String)
    case unknownStyle(Int)

    public var description: String {
        switch self {
        case .missingTensor(let name): return "missing tensor \(name)"
        case .missingSparseTensor(let name): return "missing sparse tensor \(name)"
        case .unknownStyle(let style): return "unknown style \(style)"
        }
    }
}

public final class CalligrapherModel {
    public static let stepsPerCharacter = 40

    /// The model has 80 learned styles, but many are near-duplicates or
    /// rough; calligrapher.ai's own picker exposes only these (plus random),
    /// so ours does too. The engine itself accepts any id 0-79.
    public static let exposedStyles: [Int] = [1, 2, 3, 4, 5, 6, 7, 8, 9]

    public let assets: CalligrapherAssets
    public let alphabet: [Character] = calligrapherAlphabet
    private let cell: CalligrapherCell

    public init(assets: CalligrapherAssets) throws {
        self.assets = assets
        self.cell = try CalligrapherCell(assets: assets)
    }

    public var styles: [Int] {
        Array(0 ..< assets.styleCount)
    }

    /// Encode text to model ids, wrapped in start/end markers.
    public func encode(_ text: String) -> [Int32] {
        var encoded = [Int32](repeating: 0, count: text.count + 2)
        encoded[0] = START
        for (index, character) in text.enumerated() {
            encoded[index + 1] = charToId[character] ?? UNKNOWN
        }
        encoded[text.count + 1] = END
        return encoded
    }

    public func supports(_ character: Character) -> Bool {
        charToId[character] != nil
    }

    public func writer(
        _ text: String,
        bias: Double = 0.75,
        style: Int? = nil,
        seed: UInt32 = 0
    ) throws -> CalligrapherWriter {
        try CalligrapherWriter(model: self, cell: cell, text: text, bias: bias, style: style, seed: seed)
    }

    /// Generate a full line synchronously.
    public func write(
        _ text: String,
        bias: Double = 0.75,
        style: Int? = nil,
        seed: UInt32 = 0
    ) throws -> [StrokeOffset] {
        let writer = try writer(text, bias: bias, style: style, seed: seed)
        var offsets: [StrokeOffset] = []
        while let offset = writer.step() {
            offsets.append(offset)
        }
        return offsets
    }
}

public final class CalligrapherWriter {
    public let text: String
    public let bias: Double
    public let style: Int
    public private(set) var done = false

    private let cell: CalligrapherCell
    private var rng: Rng
    private let state: CellState
    private let encoded: RawBuffer<Float>
    private let charCount: Int
    private let maxSteps: Int
    private var steps = 0

    init(model: CalligrapherModel, cell: CalligrapherCell, text: String, bias: Double, style: Int?, seed: UInt32) throws {
        self.text = text
        self.bias = bias
        self.cell = cell
        // The reference picks a random style with one uniform draw before
        // anything else; matching that keeps null-style runs reproducible.
        var rng = Rng(seed: seed)
        let chosen = style ?? Int((Double(model.assets.styleCount) * rng.uniform()).rounded(.down))
        self.rng = rng
        guard chosen >= 0, chosen < model.assets.styleCount else {
            throw CalligrapherError.unknownStyle(chosen)
        }
        self.style = chosen

        let ids = model.encode(text)
        self.charCount = ids.count
        self.encoded = RawBuffer(copying: cell.encodeText(ids))
        self.state = cell.initialState(charCount: charCount, styleIndex: chosen)
        self.maxSteps = CalligrapherModel.stepsPerCharacter * text.count
    }

    /// Advance one timestep. Returns the sampled offset, or nil once done.
    public func step() -> StrokeOffset? {
        if done { return nil }
        let termination = cell.step(state, encoded: encoded.p, charCount: charCount)
        let offset = cell.sample(state, bias: bias, rng: &rng)
        steps += 1
        if steps > maxSteps || termination > 0.5 {
            done = true
            return nil
        }
        let input = state.input.p
        input[0] = offset.dx
        input[1] = offset.dy
        input[2] = offset.pen
        return StrokeOffset(dx: Double(offset.dx), dy: Double(offset.dy), eos: offset.pen == 1)
    }
}

File: packages/ink-kotlin/ink-calligrapher/src/main/kotlin/com/trylonghand/ink/calligrapher/Engine.kt
/// Generation API over the calligrapher cell, ported from
/// packages/ink-calligrapher/src/engine.ts: encode [START, ...text, END],
/// condition on one of the 80 learned style vectors (or a seed-picked
/// random one), then autoregressively sample until the attention head
/// signals the text is exhausted or the step budget runs out. The
/// terminating step's sample is discarded, exactly like the reference.

package com.trylonghand.ink.calligrapher

import com.trylonghand.ink.core.Rng
import com.trylonghand.ink.core.StrokeOffset
import kotlin.math.floor

public sealed class CalligrapherError(message: String) : Exception(message) {
    public class MissingTensor(name: String) :
        CalligrapherError("missing tensor $name")

    public class MissingSparseTensor(name: String) :
        CalligrapherError("missing sparse tensor $name")

    public class UnknownStyle(style: Int) :
        CalligrapherError("unknown style $style")

    public class MissingBundledWeights :
        CalligrapherError("calligrapher-v1.bin is missing from the package resources")
}

public class CalligrapherModel(public val assets: CalligrapherAssets) {
    public companion object {
        public const val stepsPerCharacter: Int = 40

        /**
         * The model has 80 learned styles, but many are near-duplicates or
         * rough; calligrapher.ai's own picker exposes only these (plus random),
         * so ours does too. The engine itself accepts any id 0-79.
         */
        public val exposedStyles: List<Int> = listOf(1, 2, 3, 4, 5, 6, 7, 8, 9)
    }

    public val alphabet: List<Char> = calligrapherAlphabet
    private val cell: CalligrapherCell = CalligrapherCell(assets)

    public val styles: List<Int>
        get() = (0 until assets.styleCount).toList()

    /** Encode text to model ids, wrapped in start/end markers. */
    public fun encode(text: String): IntArray {
        val encoded = IntArray(text.length + 2)
        encoded[0] = START
        for ((index, character) in text.withIndex()) {
            encoded[index + 1] = charToId[character] ?: UNKNOWN
        }
        encoded[text.length + 1] = END
        return encoded
    }

    public fun supports(character: Char): Boolean = charToId[character] != null

    public fun writer(
        text: String,
        bias: Double = 0.75,
        style: Int? = null,
        seed: UInt = 0u,
    ): CalligrapherWriter = CalligrapherWriter(this, cell, text, bias, style, seed)

    /** Generate a full line synchronously. */
    public fun write(
        text: String,
        bias: Double = 0.75,
        style: Int? = null,
        seed: UInt = 0u,
    ): List<StrokeOffset> {
        val writer = writer(text, bias = bias, style = style, seed = seed)
        val offsets = mutableListOf<StrokeOffset>()
        while (true) {
            val offset = writer.step() ?: break
            offsets.add(offset)
        }
        return offsets
    }
}

public class CalligrapherWriter internal constructor(
    model: CalligrapherModel,
    private val cell: CalligrapherCell,
    public val text: String,
    public val bias: Double,
    style: Int?,
    seed: UInt,
) {
    public val style: Int

    public var done: Boolean = false
        private set

    private val rng: Rng
    private val state: CellState
    private val encoded: FloatArray
    private val charCount: Int
    private val maxSteps: Int
    private var steps: Int = 0

    init {
        // The reference picks a random style with one uniform draw before
        // anything else; matching that keeps null-style runs reproducible.
        val rng = Rng(seed)
        val chosen = style ?: floor(model.assets.styleCount.toDouble() * rng.uniform()).toInt()
        this.rng = rng
        if (chosen < 0 || chosen >= model.assets.styleCount) {
            throw CalligrapherError.UnknownStyle(chosen)
        }
        this.style = chosen

        val ids = model.encode(text)
        charCount = ids.size
        encoded = cell.encodeText(ids)
        state = cell.initialState(charCount = charCount, styleIndex = chosen)
        maxSteps = CalligrapherModel.stepsPerCharacter * text.length
    }

    /** Advance one timestep. Returns the sampled offset, or null once done. */
    public fun step(): StrokeOffset? {
        if (done) return null
        val termination = cell.step(state, encoded = encoded, charCount = charCount)
        val offset = cell.sample(state, bias = bias, rng = rng)
        steps += 1
        if (steps > maxSteps || termination > 0.5f) {
            done = true
            return null
        }
        val input = state.input
        input[0] = offset.dx
        input[1] = offset.dy
        input[2] = offset.pen
        return StrokeOffset(dx = offset.dx.toDouble(), dy = offset.dy.toDouble(), eos = offset.pen == 1f)
    }
}
